### ISSUE TITLE
Unit test failure

### ISSUE THREAD

**Issue by mpaszkow** [carlobaldassi/GaussDCA.jl#4]
#### Description
Unit testing of the package fails. Whereas the first part of the `test1` unit gives an unexpected result (0.8330339 _vs._ 0.3001522), all other invocations of `gDCA` fail with an error:
```
ERROR: PosDefException: matrix is not positive definite; Cholesky factorization failed.
```

#### To Reproduce
Enter `pkg` mode by pressing `]` 
```
(v1.1) pkg> test GaussDCA
```
Output:
```
   Testing GaussDCA
 Resolving package versions...
    Status `/tmp/tmpAji1Kz/Manifest.toml`
  [34da2185] Compat v2.1.0
  [a0c94c4b] FastaIO v0.5.0
  [92fee26a] GZip v0.5.0
  [92cae0f8] GaussDCA v0.0.0 [`~/.julia/dev/GaussDCA`]
  [2a0f44e3] Base64  [`@stdlib/Base64`]
  [ade2ca70] Dates  [`@stdlib/Dates`]
  [8bb1440f] DelimitedFiles  [`@stdlib/DelimitedFiles`]
  [8ba89e20] Distributed  [`@stdlib/Distributed`]
  [b77e0a4c] InteractiveUtils  [`@stdlib/InteractiveUtils`]
  [76f85450] LibGit2  [`@stdlib/LibGit2`]
  [8f399da3] Libdl  [`@stdlib/Libdl`]
  [37e2e46d] LinearAlgebra  [`@stdlib/LinearAlgebra`]
  [56ddb016] Logging  [`@stdlib/Logging`]
  [d6f4376e] Markdown  [`@stdlib/Markdown`]
  [a63ad114] Mmap  [`@stdlib/Mmap`]
  [44cfe95a] Pkg  [`@stdlib/Pkg`]
  [de0858da] Printf  [`@stdlib/Printf`]
  [3fa0cd96] REPL  [`@stdlib/REPL`]
  [9a3f8284] Random  [`@stdlib/Random`]
  [ea8e919c] SHA  [`@stdlib/SHA`]
  [9e88b42a] Serialization  [`@stdlib/Serialization`]
  [1a1011a3] SharedArrays  [`@stdlib/SharedArrays`]
  [6462fe0b] Sockets  [`@stdlib/Sockets`]
  [2f01184e] SparseArrays  [`@stdlib/SparseArrays`]
  [10745b16] Statistics  [`@stdlib/Statistics`]
  [8dfed614] Test  [`@stdlib/Test`]
  [cf7118a7] UUIDs  [`@stdlib/UUIDs`]
  [4ec0a83e] Unicode  [`@stdlib/Unicode`]
[ Info: Running Main.GaussDCATests.test1
theta = 0.3633365278439081 threshold = 19.0
M = 106 N = 53 Meff = 91.99999999999997
Test Failed at ~/.julia/dev/GaussDCA/test/runtests.jl:51
  Expression: d1[k] ≈ d2[k]
   Evaluated: 0.8330339 ≈ 0.3001522
ERROR: LoadError: There was an error during testing
in expression starting at ~/.julia/dev/GaussDCA/test/runtests.jl:93
ERROR: Package GaussDCA errored during testing
```
Since not all tests are executed, I run the other two individually:
```
julia> test2()
removing duplicate sequences... done: 97 -> 94
theta = 0.3681218129968263 threshold = 147.0
M = 94 N = 400 Meff = 25.8138802992673
ERROR: PosDefException: matrix is not positive definite; Cholesky factorization failed.
Stacktrace:
 [1] checkpositivedefinite at /build/julia/src/julia-1.1.1/usr/share/julia/stdlib/v1.1/LinearAlgebra/src/factorization.jl:11 [inlined]
 [2] #cholesky!#96(::Bool, ::Function, ::LinearAlgebra.Hermitian{Float64,Array{Float64,2}}, ::Val{false}) at /build/julia/src/julia-1.1.1/usr/share/julia/stdlib/v1.1/LinearAlgebra/src/cholesky.jl:153
 [3] #cholesky! at ./none:0 [inlined]
 [4] #cholesky!#97(::Bool, ::Function, ::Array{Float64,2}, ::Val{false}) at /build/julia/src/julia-1.1.1/usr/share/julia/stdlib/v1.1/LinearAlgebra/src/cholesky.jl:185
 [5] #cholesky! at ./none:0 [inlined] (repeats 2 times)
 [6] #cholesky#101 at /build/julia/src/julia-1.1.1/usr/share/julia/stdlib/v1.1/LinearAlgebra/src/cholesky.jl:275 [inlined]
 [7] cholesky at /build/julia/src/julia-1.1.1/usr/share/julia/stdlib/v1.1/LinearAlgebra/src/cholesky.jl:275 [inlined] (repeats 2 times)
 [8] #gDCA#1(::Float64, ::Symbol, ::Float64, ::Symbol, ::Int64, ::Bool, ::Function, ::String) at ~/.julia/dev/GaussDCA/src/GaussDCA.jl:48
 [9] (::getfield(GaussDCA, Symbol("#kw##gDCA")))(::NamedTuple{(:pseudocount, :score, :remove_dups),Tuple{Float64,Symbol,Bool}}, ::typeof(gDCA), ::String) at ./none:0
 [10] test2() at ./REPL[57]:2
 [11] top-level scope at none:0
```
```
julia> test3()
removing duplicate sequences... done: 106 -> 98
GaussDCA: using slower fallbacks
theta = 0.3626005800258088 threshold = 19.0
M = 98 N = 53 Meff = 92.0
ERROR: PosDefException: matrix is not positive definite; Cholesky factorization failed.
Stacktrace:
 [1] checkpositivedefinite at /build/julia/src/julia-1.1.1/usr/share/julia/stdlib/v1.1/LinearAlgebra/src/factorization.jl:11 [inlined]
 [2] #cholesky!#96(::Bool, ::Function, ::LinearAlgebra.Hermitian{Float64,Array{Float64,2}}, ::Val{false}) at /build/julia/src/julia-1.1.1/usr/share/julia/stdlib/v1.1/LinearAlgebra/src/cholesky.jl:153
 [3] #cholesky! at ./none:0 [inlined]
 [4] #cholesky!#97(::Bool, ::Function, ::Array{Float64,2}, ::Val{false}) at /build/julia/src/julia-1.1.1/usr/share/julia/stdlib/v1.1/LinearAlgebra/src/cholesky.jl:185
 [5] #cholesky! at ./none:0 [inlined] (repeats 2 times)
 [6] #cholesky#101 at /build/julia/src/julia-1.1.1/usr/share/julia/stdlib/v1.1/LinearAlgebra/src/cholesky.jl:275 [inlined]
 [7] cholesky at /build/julia/src/julia-1.1.1/usr/share/julia/stdlib/v1.1/LinearAlgebra/src/cholesky.jl:275 [inlined] (repeats 2 times)
 [8] #gDCA#1(::Float64, ::Symbol, ::Float64, ::Symbol, ::Int64, ::Bool, ::Function, ::String) at /home/macpasz/.julia/dev/GaussDCA/src/GaussDCA.jl:48
 [9] #gDCA at ./none:0 [inlined]
 [10] test3() at ./REPL[59]:3
 [11] top-level scope at none:0
```

I am using Julia 1.1.1 with OpenBLAS 0.3.6

**Comment by carlobaldassi** [comment]
It works for fine me both on julia 1.0 and on 1.1.2-pre. It also worked fine in the CI tests.

However, I think I'm on OpenBLAS 0.3.2 on julia 1.0 and on OpenBLAS 0.3.3 on julia 1.1, which are the versions that come with the julia source. How did you get version 0.3.6, and how did you check its version from julia?

I get the same outputs in tests 2 and 3, but no complaints from the cholesky decomposition. Since that seems to be the problem, it may be about BLAS not working properly? Does cholesky decomposition work for you otherwise? For example something like

```
julia> using LinearAlgebra

julia> X = rand(500, 100);

julia> C = cholesky(X'X);

julia> U = C.U;

julia> @assert maximum(abs.(U'U - X'X)) < 1e-12
```

**Comment by mpaszkow** [comment]
I haven't realized that julia comes with its own BLAS libraries - I reported a version of my stand-alone OpenBLAS package, which probably is not used by julia. The simple test code that you suggested fails to execute on my installation:
```
julia> using LinearAlgebra
julia> X = rand(500, 100);
julia> C = cholesky(X'X);
ERROR: PosDefException: matrix is not positive definite; Cholesky factorization failed.
Stacktrace:
 [1] checkpositivedefinite at /build/julia/src/julia-1.1.1/usr/share/julia/stdlib/v1.1/LinearAlgebra/src/factorization.jl:11 [inlined]
 [2] #cholesky!#96(::Bool, ::Function, ::Hermitian{Float64,Array{Float64,2}}, ::Val{false}) at /build/julia/src/julia-1.1.1/usr/share/julia/stdlib/v1.1/LinearAlgebra/src/cholesky.jl:153
 [3] #cholesky! at ./none:0 [inlined]
 [4] #cholesky!#97(::Bool, ::Function, ::Array{Float64,2}, ::Val{false}) at /build/julia/src/julia-1.1.1/usr/share/julia/stdlib/v1.1/LinearAlgebra/src/cholesky.jl:185
 [5] #cholesky#101 at ./none:0 [inlined]
 [6] cholesky at /build/julia/src/julia-1.1.1/usr/share/julia/stdlib/v1.1/LinearAlgebra/src/cholesky.jl:275 [inlined] (repeats 2 times)
 [7] top-level scope at none:0
```
Judging from various comments on the Internet, it seems to be a general julia issue. I will try installing the latest version from git and let's see if it helps.

**Comment by mpaszkow** [comment]
I solved the issue by compiling and installing the latest git version of the OpenBLAS library. I'm using Arch Linux and, in this distribution, julia is compiled to use system-wide libraries instead of its own versions. Apparently, the _openblas_ package that I had installed from the official repository (version 0.3.6-1) was not working well. Maybe there is also some hardware-related incompatibility that would not manifest on other CPUs.

Anyway, just as an information for other Arch Linux users that may run into a similar issue, installing the [openblas-lapack](https://aur.archlinux.org/packages/openblas-lapack/) package from AUR solved the problem.
